tss-esapi/examples/duplication.rs

```diff
@@ -0,0 +1,427 @@
+                policy_auth_session_attributes_mask,
+            )
+            .expect("tr_sess_set_attributes call failed");
+
+            let policy_session = PolicySession::try_from(trial_session)
+                .expect("Failed to convert auth session into policy session");
+
+            ctx.policy_duplication_select(
+                policy_session,
+                Vec::<u8>::new().try_into().unwrap(),
+                target_parent_name.clone(),
+                false,
+            )
+            .expect("Policy duplication select");
+
+            ctx.policy_get_digest(policy_session)
+        })
+        .unwrap();
+
+    // Create the storage key on TPM-1, using the policy digest we just created.
+    let object_attributes = ObjectAttributesBuilder::new()
+        // For a key to be duplicated, it must have fixed TPM and parent as false.
+        .with_fixed_tpm(false)
+        .with_fixed_parent(false)
+        .with_sensitive_data_origin(true)
+        .with_user_with_auth(true)
+        .with_decrypt(true)
+        .with_sign_encrypt(true)
+        .with_restricted(false)
+        .build()
+        .expect("Attributes to be valid");
+
+    let storage_public = PublicBuilder::new()
+        .with_public_algorithm(PublicAlgorithm::Ecc)
+        .with_name_hashing_algorithm(HashingAlgorithm::Sha256)
+        .with_object_attributes(object_attributes)
+        // Use policy digest computed using the trial session
+        .with_auth_policy(policy_digest)
+        .with_ecc_parameters(
+            PublicEccParametersBuilder::new()
+                .with_ecc_scheme(EccScheme::Null)
+                .with_curve(EccCurve::NistP256)
+                .with_is_signing_key(false)
+                .with_is_decryption_key(true)
+                .with_restricted(false)
+                .with_key_derivation_function_scheme(KeyDerivationFunctionScheme::Null)
+                .build()
+                .expect("Params to be valid"),
+        )
+        .with_ecc_unique_identifier(EccPoint::default())
+        .build()
+        .expect("storage public to be valid");
+
+    let storage_key = context_1
+        .execute_with_nullauth_session(|ctx| {
+            ctx.create(
+                primary_key_1.key_handle,
+                storage_public,
+                None,
+                None,
+                None,
+                None,
+            )
+        })
+        .unwrap();
+
+    // Load the key.
+    let loaded_storage_key = context_1
+        .execute_with_nullauth_session(|ctx| {
+            ctx.load(
+                primary_key_1.key_handle,
+                storage_key.out_private.clone(),
+                storage_key.out_public.clone(),
+            )
+        })
+        .unwrap();
+
+    // Now we can create a child key that we will be able to move along with the parent.
+    //
+    // ALWAYS FAILS!!!
+    // 0x0000018a TpmFormatOneResponseCode { error_number: Type, argument_number: Handle(1) }
+    /*
```

Comment:
Is this the actual error you get when you try to display it? If you use master you should get a way more comprehensive error then that if you just take the exception and print it. It should tell you something about which layer that caused the error and so on.

Reply:
When displayed it's `⚠️  TSS Layer: TPM, Code: 0x0000018A, Message: The type of the value is not appropriate for the use (associated with handle number 1).`. Which is still pretty vague. I'm guessing it's something about the storage key and it's attributes that is preventing it from having a child key this type?

Reply:
Yes, something about how that key is created prevents it from being a storage key.

Reply:
Yeah for sure. But whats weird is this is using https://docs.rs/tss-esapi/7.4.0/tss_esapi/structures/struct.PublicEccParametersBuilder.html#method.new_restricted_decryption_key which claims to be valid for a storage key. So clearly there is a bug here in these params and I lack the experience and skills to solve it :(